fdbserver/RESTKmsConnector.actor.cpp

```diff
@@ -0,0 +1,1143 @@
+	//        "encryptDomainId" : <domainId>
+	//     },
+	//     {
+	//         ....
+	//	   }
+	//   ],
+	//   "validationTokens" = [
+	//     {
+	//        "name" : <name>,
+	//        "value": <value>
+	//     },
+	//     {
+	//         ....
+	//     }
+	//   ]
+	//   "refreshKmsUrls" = 1/0
+	// }
+
+	rapidjson::Document doc;
+	doc.SetObject();
+
+	// Append "keyIdDetails" as json array
+	rapidjson::Value keyIdDetails(rapidjson::kArrayType);
+	for (const auto& detail : req.encryptKeyIds) {
+		rapidjson::Value keyIdDetail(rapidjson::kObjectType);
+
+		// Add "baseCipherId"
+		rapidjson::Value key(BASE_CIPHER_ID_TAG, doc.GetAllocator());
+		rapidjson::Value baseKeyId;
+		baseKeyId.SetUint64(detail.first);
+		keyIdDetail.AddMember(key, baseKeyId, doc.GetAllocator());
+
+		// Add "encryptDomainId"
+		key.SetString(ENCRYPT_DOMAIN_ID_TAG, doc.GetAllocator());
+		rapidjson::Value domainId;
+		domainId.SetInt64(detail.second);
+		keyIdDetail.AddMember(key, domainId, doc.GetAllocator());
+
+		// push above object to the array
+		keyIdDetails.PushBack(keyIdDetail, doc.GetAllocator());
+	}
+	rapidjson::Value memberKey(CIPHER_KEY_DETAILS_TAG, doc.GetAllocator());
+	doc.AddMember(memberKey, keyIdDetails, doc.GetAllocator());
+
+	// Append "validationTokens" as json array
+	addValidationTokensSectionToJsonDoc(ctx, doc);
+
+	// Append "refreshKmsUrls"
+	addRefreshKmsUrlsSectionToJsonDoc(ctx, doc, refreshKmsUrls);
+
+	// Serialize json to string
+	rapidjson::StringBuffer sb;
+	rapidjson::Writer<rapidjson::StringBuffer> writer(sb);
+	doc.Accept(writer);
+	outJsonStr.resize(sb.GetSize());
+	memcpy(outJsonStr.data(), sb.GetString(), sb.GetSize());
+}
+
+ACTOR Future<KmsConnLookupEKsByKeyIdsRep> fetchEncryptionKeyByKeyId(Reference<RESTKmsConnectorCtx> ctx,
+                                                                    KmsConnLookupEKsByKeyIdsReq req) {
+	state Reference<HTTP::Response> resp;
+	state KmsConnLookupEKsByKeyIdsRep reply;
+	state bool refreshKmsUrls = shouldRefreshKmsUrls(ctx);
+	state std::string requestBody;
+
+	populateGetEncryptKeysByKeyIdsRequestBody(ctx, req, refreshKmsUrls, requestBody);
+
+	// Follow 2-phase scheme:
+	// Phase-1: Attempt to fetch encryption keys by reaching out to cached KmsUrls in the order of
+	//          past success requests success counts.
+	// Phase-2: For some reason if none of the cached KmsUrls worked, re-discover the KmsUrls and
+	//          repeat phase-1.
+
+	state int pass = 1;
+	for (; pass <= 2; pass++) {
+		state std::stack<std::shared_ptr<KmsUrlCtx>> tempStack;
+
+		// Iterate over Kms URLs
+		while (!ctx->kmsUrlHeap.empty()) {
+			state std::shared_ptr<KmsUrlCtx> curUrl = ctx->kmsUrlHeap.top();
+			ctx->kmsUrlHeap.pop();
+			tempStack.push(curUrl);
+
+			try {
+				std::string kmsEncryptionFullUrl = getEncryptionFullUrl(curUrl->url);
+				TraceEvent("FetchEncryptionKeyByKeyId_Start", ctx->uid)
+				    .detail("KmsEncryptionFullUrl", Traceable<std::string>::toString(kmsEncryptionFullUrl));
```

Comment:
See above, I think here you can just pass `kmsEncryptionFullUrl` and you don't need `Traceable`